core/src/main/java/org/elasticsearch/common/util/NamedEnum.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util;
+
+import java.util.EnumSet;
+
+import static org.elasticsearch.common.util.Preconditions.checkNotNull;
+
+/**
+ * An interface that enums should implement if retrieving the enum by its
+ * assigned name is a desired operation.
+ */
+public interface NamedEnum {
+
+    /**
+     * The name by which the enumeration is identified.
+     * @return The name of the enum
+     */
+    String getName();
+
+    /**
+     * Get the correct enum value for the given enum type and name.  The enum
+     * type must implement the {@link NamedEnum} interface for this to work.
+     *
+     * @param cls The class of the enum
+     * @param name The name of the enum within the given type
+     * @return The enum value matching the given name
+     * @throws IllegalArgumentException if the input is invalid or an enum value with the given name could not be found
+     */
+    static <E extends Enum<E>> E fromString(final Class<E> cls, final String name) {
+        if (name == null) {
+            return null;
+        }
+        checkNotNull(cls, "class");
+        if (!NamedEnum.class.isAssignableFrom(cls)) {
```

Comment:
Just curious if you have any experience about performance with this checks, and the `allOf` call below. Maybe negligable, but since we will call this in serialization if we use the enum names there, it might be called a lot.